system/inc/system_sleep.h

```diff
@@ -41,12 +45,147 @@ typedef enum System_Sleep_Flag
+    }
+
+    SystemSleepConfiguration& gpio(pin_t pin, InterruptMode mode) {
+        if (!valid_) {
+            return *this;
+        }
+        // Check if this pin has been featured.
+        auto wakeup = wakeupSourceFeatured(HAL_WAKEUP_SOURCE_TYPE_GPIO);
+        while (wakeup) {
+            auto gpioWakeup = reinterpret_cast<hal_wakeup_source_gpio_t*>(wakeup);
+            if (gpioWakeup->pin == pin) {
+                gpioWakeup->mode = mode;
+                return *this;
+            }
+            wakeup = wakeupSourceFeatured(HAL_WAKEUP_SOURCE_TYPE_GPIO, wakeup->next);
+        }
+        // Otherwise, configure this pin as wakeup source.
+        auto wakeupSource = new(std::nothrow) hal_wakeup_source_gpio_t();
+        if (!wakeupSource) {
+            valid_ = false;
+            return *this;
+        }
+        wakeupSource->base.size = sizeof(hal_wakeup_source_gpio_t);
+        wakeupSource->base.version = HAL_SLEEP_VERSION;
+        wakeupSource->base.type = HAL_WAKEUP_SOURCE_TYPE_GPIO;
+        wakeupSource->base.next = config_.wakeup_sources;
+        wakeupSource->pin = pin;
+        wakeupSource->mode = mode;
+        config_.wakeup_sources = reinterpret_cast<hal_wakeup_source_base_t*>(wakeupSource);
+        return *this;
+    }
+
+    SystemSleepConfiguration& duration(system_tick_t ms) {
+        if (!valid_) {
+            return *this;
+        }
+        // Check if RTC has been configured as wakeup source.
+        auto wakeup = wakeupSourceFeatured(HAL_WAKEUP_SOURCE_TYPE_RTC);
+        if (wakeup) {
+            reinterpret_cast<hal_wakeup_source_rtc_t*>(wakeup)->ms = ms;
+            return *this;
+        }
+        // Otherwise, configure RTC as wakeup source.
+        auto wakeupSource = new(std::nothrow) hal_wakeup_source_rtc_t();
+        if (!wakeupSource) {
+            valid_ = false;
+            return *this;
+        }
+        wakeupSource->base.size = sizeof(hal_wakeup_source_rtc_t);
+        wakeupSource->base.version = HAL_SLEEP_VERSION;
+        wakeupSource->base.type = HAL_WAKEUP_SOURCE_TYPE_RTC;
+        wakeupSource->base.next = config_.wakeup_sources;
+        wakeupSource->ms = ms;
+        config_.wakeup_sources = reinterpret_cast<hal_wakeup_source_base_t*>(wakeupSource);
+        return *this;
+    }
+
+    SystemSleepConfiguration& duration(std::chrono::milliseconds ms) {
+        return duration(ms.count());
+    }
+
+private:
+    hal_sleep_config_t config_;
+    bool valid_;
+};
+
+
 /**
  * @param param A SystemSleepNetwork enum cast as an integer.
  */
 int system_sleep(Spark_Sleep_TypeDef mode, long seconds, uint32_t param, void* reserved);
 int system_sleep_pin(uint16_t pin, uint16_t mode, long seconds, uint32_t param, void* reserved);
 int system_sleep_pins(const uint16_t* pins, size_t pins_count, const InterruptMode* modes, size_t modes_count, long seconds, uint32_t param, void* reserved);
+int system_sleep_ext(const SystemSleepConfiguration& config, void* reserved);
```

Comment:
`const SystemSleepConfiguration& config` -> `const SystemSleepConfiguration* config`

Reply:
To avoid invoking the copy constructor?

Reply:
Nope, you simply shouldn't use references in dynalib APIs.

Reply:
Ah 👍  ! I was thinking the system layer is compiled along with the wiring layer.